experimental/apis/data/v0alpha1/query.go

```diff
@@ -0,0 +1,419 @@
+				}
+			}
+
+		case "datasourceId":
+			g.DatasourceID, err = iter.ReadInt64()
+		case "queryType":
+			g.QueryType, err = iter.ReadString()
+		case "maxDataPoints":
+			g.MaxDataPoints, err = iter.ReadInt64()
+		case "intervalMs":
+			g.IntervalMS, err = iter.ReadFloat64()
+		case "hide":
+			g.Hide, err = iter.ReadBool()
+		default:
+			err = processUnknownKey(field, iter)
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+// CommonQueryProperties are properties that can be added to all queries.
+// These properties live in the same JSON level as datasource specific properties,
+// so care must be taken to ensure they do not overlap
+type CommonQueryProperties struct {
+	// RefID is the unique identifier of the query, set by the frontend call.
+	RefID string `json:"refId,omitempty"`
+
+	// Optionally define expected query result behavior
+	ResultAssertions *ResultAssertions `json:"resultAssertions,omitempty"`
+
+	// TimeRange represents the query range
+	// NOTE: unlike generic /ds/query, we can now send explicit time values in each query
+	// NOTE: the values for timeRange are not saved in a dashboard, they are constructed on the fly
+	TimeRange *TimeRange `json:"timeRange,omitempty"`
+
+	// The datasource
+	Datasource *DataSourceRef `json:"datasource,omitempty"`
+
+	// Deprecated -- use datasource ref instead
+	DatasourceID int64 `json:"datasourceId,omitempty"`
+
+	// QueryType is an optional identifier for the type of query.
+	// It can be used to distinguish different types of queries.
+	QueryType string `json:"queryType,omitempty"`
+
+	// MaxDataPoints is the maximum number of data points that should be returned from a time series query.
+	// NOTE: the values for maxDataPoints is not saved in the query model.  It is typically calculated
+	// from the number of pixels visible in a visualization
+	MaxDataPoints int64 `json:"maxDataPoints,omitempty"`
+
+	// Interval is the suggested duration between time points in a time series query.
+	// NOTE: the values for intervalMs is not saved in the query model.  It is typically calculated
+	// from the interval required to fill a pixels in the visualization
+	IntervalMS float64 `json:"intervalMs,omitempty"`
+
+	// true if query is disabled (ie should not be returned to the dashboard)
+	// NOTE: this does not always imply that the query should not be executed since
+	// the results from a hidden query may be used as the input to other queries (SSE etc)
+	Hide bool `json:"hide,omitempty"`
+}
+
+type DataSourceRef struct {
+	// The datasource plugin type
+	Type string `json:"type"`
+
+	// Datasource UID
+	UID string `json:"uid,omitempty"`
+
+	// ?? the datasource API version?  (just version, not the group? type | apiVersion?)
+}
+
+// TimeRange represents a time range for a query and is a property of DataQuery.
+type TimeRange struct {
+	// From is the start time of the query.
+	From string `json:"from" jsonschema:"example=now-1h,default=now-6h"`
```

Comment:
- From and Two should be typed beyond being a string so we can have valid and valid values for what they are.
- If relative times are supported, there should also be an (optional) Timestamp to indicate an absolute value for "now".
- We should have methods or function that return time.Time for these, so that does not vary between receivers.

Reply:
If not, then things like following are valid, and if they don't work it is a bug ;-)

![image](https://github.com/grafana/grafana-plugin-sdk-go/assets/1692624/9cd175b3-e329-417e-a1a5-207bc9591351)